File: nodejs-whisper-transcriber/src/vad.js
const mic = require('mic');
const { spawn } = require('child_process');
const fs = require('fs');
const path = require('path');

class VAD {
    constructor() {
        this.isRecording = false;
        this.micInstance = null;
        this.micInputStream = null;
        this.currentAudioFile = null;
        this.silenceTimeout = null;
        this.audioCounter = 0;
        this.tempDir = path.join(__dirname, '../data/temp');
        
        // Configurações VAD
        this.silenceThreshold = 1000; // ms de silêncio para parar
        this.minRecordingTime = 500; // ms mínimo de gravação
        this.recordingStartTime = null;
        
        this.ensureTempDir();
    }

    ensureTempDir() {
        if (!fs.existsSync(this.tempDir)) {
            fs.mkdirSync(this.tempDir, { recursive: true });
        }
    }

    startListening() {
        console.log('Iniciando escuta por voz...');
        
        // Configuração do microfone
        this.micInstance = mic({
            rate: '16000',
            channels: '1',
            debug: false,
            exitOnSilence: 0
        });

        this.micInputStream = this.micInstance.getAudioStream();
        
        // Detectar início de fala
        this.micInputStream.on('data', (data) => {
            this.handleAudioData(data);
        });

        this.micInputStream.on('error', (err) => {
            console.error('Erro no microfone:', err);
        });

        this.micInstance.start();
        console.log('Microfone ativo. Aguardando voz...');
    }

    handleAudioData(data) {
        const audioLevel = this.calculateAudioLevel(data);
        
        // Detectar início de fala (threshold simples)
        if (audioLevel > 5000 && !this.isRecording) {
            this.startRecording();
        }
        
        // Durante a gravação, resetar timeout de silêncio
        if (this.isRecording) {
            if (audioLevel > 3000) {
                this.resetSilenceTimeout();
            }
            this.writeAudioData(data);
        }
    }

    calculateAudioLevel(buffer) {
        let sum = 0;
        for (let i = 0; i < buffer.length; i += 2) {
            const sample = buffer.readInt16LE(i);
            sum += Math.abs(sample);
        }
        return sum / (buffer.length / 2);
    }

    startRecording() {
        if (this.isRecording) return;
        
        this.isRecording = true;
        this.recordingStartTime = Date.now();
        this.audioCounter++;
        
        const filename = `audio_${String(this.audioCounter).padStart(4, '0')}.wav`;
        this.currentAudioFile = path.join(this.tempDir, filename);
        
        console.log(`Iniciando gravação: ${filename}`);
        
        // Criar arquivo WAV com cabeçalho
        this.createWavFile(this.currentAudioFile);
        
        this.resetSilenceTimeout();
    }

    writeAudioData(data) {
        if (!this.isRecording || !this.currentAudioFile) return;
        
        fs.appendFileSync(this.currentAudioFile, data);
    }

    resetSilenceTimeout() {
        if (this.silenceTimeout) {
            clearTimeout(this.silenceTimeout);
        }
        
        this.silenceTimeout = setTimeout(() => {
            this.stopRecording();
        }, this.silenceThreshold);
    }

    stopRecording() {
        if (!this.isRecording) return;
        
        const recordingDuration = Date.now() - this.recordingStartTime;
        
        // Verificar duração mínima
        if (recordingDuration < this.minRecordingTime) {
            console.log('Gravação muito curta, descartando...');
            this.discardCurrentRecording();
            return;
        }
        
        this.isRecording = false;
        console.log(`Parando gravação: ${path.basename(this.currentAudioFile)}`);
        
        // Finalizar arquivo WAV
        this.finalizeWavFile(this.currentAudioFile);
        
        // Notificar que há um arquivo pronto para transcrição
        if (this.onAudioReady) {
            this.onAudioReady(this.currentAudioFile);
        }
        
        this.currentAudioFile = null;
        
        if (this.silenceTimeout) {
            clearTimeout(this.silenceTimeout);
            this.silenceTimeout = null;
        }
    }

    discardCurrentRecording() {
        if (this.currentAudioFile && fs.existsSync(this.currentAudioFile)) {
            fs.unlinkSync(this.currentAudioFile);
        }
        this.isRecording = false;
        this.currentAudioFile = null;
    }

    createWavFile(filename) {
        const sampleRate = 16000;
        const channels = 1;
        const bitsPerSample = 16;
        
        const header = Buffer.alloc(44);
        
        // RIFF header
        header.write('RIFF', 0);
        header.writeUInt32LE(36, 4); // File size - 8 (will be updated later)
        header.write('WAVE', 8);
        
        // fmt chunk
        header.write('fmt ', 12);
        header.writeUInt32LE(16, 16); // Subchunk1Size
        header.writeUInt16LE(1, 20); // AudioFormat (PCM)
        header.writeUInt16LE(channels, 22);
        header.writeUInt32LE(sampleRate, 24);
        header.writeUInt32LE(sampleRate * channels * bitsPerSample / 8, 28); // ByteRate
        header.writeUInt16LE(channels * bitsPerSample / 8, 32); // BlockAlign
        header.writeUInt16LE(bitsPerSample, 34);
        
        // data chunk
        header.write('data', 36);
        header.writeUInt32LE(0, 40); // Subchunk2Size (will be updated later)
        
        fs.writeFileSync(filename, header);
    }

    finalizeWavFile(filename) {
        const stats = fs.statSync(filename);
        const fileSize = stats.size;
        const dataSize = fileSize - 44;
        
        const fd = fs.openSync(filename, 'r+');
        
        // Update file size
        const fileSizeBuffer = Buffer.alloc(4);
        fileSizeBuffer.writeUInt32LE(fileSize - 8, 0);
        fs.writeSync(fd, fileSizeBuffer, 0, 4, 4);
        
        // Update data size
        const dataSizeBuffer = Buffer.alloc(4);
        dataSizeBuffer.writeUInt32LE(dataSize, 0);
        fs.writeSync(fd, dataSizeBuffer, 0, 4, 40);
        
        fs.closeSync(fd);
    }

    stop() {
        if (this.micInstance) {
            this.micInstance.stop();
        }
        if (this.silenceTimeout) {
            clearTimeout(this.silenceTimeout);
        }
        console.log('VAD parado.');
    }
}

module.exports = VAD;

File: python-whisper-transcriber/src/vad.py
import pyaudio
import webrtcvad
import wave
import time
import os
import threading
import queue
from collections import deque
import logging

logger = logging.getLogger(__name__)

class VAD:
    def __init__(self, config=None):
        self.config = config or {}
        
        # Configurações de áudio
        self.sample_rate = self.config.get('sample_rate', 16000)
        self.channels = 1
        self.chunk_duration_ms = self.config.get('chunk_duration_ms', 30)  # 30ms chunks para webrtcvad
        self.chunk_size = int(self.sample_rate * self.chunk_duration_ms / 1000)
        
        # Configurações VAD
        self.vad_aggressiveness = self.config.get('vad_aggressiveness', 2)  # 0-3
        self.silence_duration_ms = self.config.get('silence_duration_ms', 1000)
        self.min_recording_duration_ms = self.config.get('min_recording_duration_ms', 500)
        
        # Estado
        self.is_recording = False
        self.is_listening = False
        self.audio_counter = 0
        self.current_audio_file = None
        self.recording_start_time = None
        
        # Buffers
        self.silence_chunks = int(self.silence_duration_ms / self.chunk_duration_ms)
        self.voice_buffer = deque(maxlen=self.silence_chunks)
        self.audio_buffer = deque()
        
        # Threading
        self.audio_queue = queue.Queue()
        self.stop_flag = threading.Event()
        self.recording_thread = None
        
        # Callbacks
        self.on_audio_ready = None
        
        # Diretórios
        self.temp_dir = self.config.get('temp_dir', './data/temp')
        os.makedirs(self.temp_dir, exist_ok=True)
        
        # Inicializar componentes
        self.vad = webrtcvad.Vad(self.vad_aggressiveness)
        self.audio = None
        self.stream = None

    def setup_audio(self):
        """Configurar interface de áudio"""
        try:
            self.audio = pyaudio.PyAudio()
            
            # Encontrar dispositivo de entrada padrão
            device_info = self.audio.get_default_input_device_info()
            logger.info(f"Usando dispositivo: {device_info['name']}")
            
            self.stream = self.audio.open(
                format=pyaudio.paInt16,
                channels=self.channels,
                rate=self.sample_rate,
                input=True,
                frames_per_buffer=self.chunk_size,
                input_device_index=device_info['index']
            )
            
            return True
            
        except Exception as e:
            logger.error(f"Erro ao configurar áudio: {e}")
            return False

    def start_listening(self):
        """Iniciar escuta contínua"""
        if self.is_listening:
            logger.warning("VAD já está escutando")
            return
            
        if not self.setup_audio():
            logger.error("Falha ao configurar áudio")
            return False
            
        self.is_listening = True
        self.stop_flag.clear()
        
        # Iniciar thread de gravação
        self.recording_thread = threading.Thread(target=self._recording_loop, daemon=True)
        self.recording_thread.start()
        
        logger.info("VAD iniciado - aguardando voz...")
        return True

    def _recording_loop(self):
        """Loop principal de gravação"""
        try:
            while not self.stop_flag.is_set():
                # Ler chunk de áudio
                try:
                    data = self.stream.read(self.chunk_size, exception_on_overflow=False)
                    self._process_audio_chunk(data)
                except Exception as e:
                    logger.error(f"Erro na leitura de áudio: {e}")
                    break
                    
        except Exception as e:
            logger.error(f"Erro no loop de gravação: {e}")
        finally:
            self._cleanup_audio()

    def _process_audio_chunk(self, audio_chunk):
        """Processar chunk de áudio para VAD"""
        # Verificar se é voz
        is_speech = self.vad.is_speech(audio_chunk, self.sample_rate)
        
        if is_speech:
            logger.debug("Voz detectada")
            
            # Se não estava gravando, iniciar
            if not self.is_recording:
                self._start_recording()
                
                # Adicionar chunks de voz anteriores (pre-buffer)
                for buffered_chunk in self.voice_buffer:
                    self.audio_buffer.append(buffered_chunk)
                    
            # Adicionar chunk atual
            self.audio_buffer.append(audio_chunk)
            
            # Resetar contador de silêncio
            self.voice_buffer.clear()
            
        else:
            # Silêncio detectado
            if self.is_recording:
                # Adicionar ao buffer de silêncio
                self.voice_buffer.append(audio_chunk)
                self.audio_buffer.append(audio_chunk)
                
                # Se buffer de silêncio estiver cheio, parar gravação
                if len(self.voice_buffer) >= self.silence_chunks:
                    self._stop_recording()
            else:
                # Manter buffer circular para pre-recording
                self.voice_buffer.append(audio_chunk)

    def _start_recording(self):
        """Iniciar nova gravação"""
        if self.is_recording:
            return
            
        self.is_recording = True
        self.recording_start_time = time.time()
        self.audio_counter += 1
        
        filename = f"audio_{self.audio_counter:04d}.wav"
        self.current_audio_file = os.path.join(self.temp_dir, filename)
        
        # Limpar buffer de áudio
        self.audio_buffer.clear()
        
        logger.info(f"Iniciando gravação: {filename}")

    def _stop_recording(self):
        """Parar gravação atual"""
        if not self.is_recording:
            return
            
        recording_duration = (time.time() - self.recording_start_time) * 1000
        
        # Verificar duração mínima
        if recording_duration < self.min_recording_duration_ms:
            logger.info(f"Gravação muito curta ({recording_duration:.0f}ms), descartando")
            self._discard_recording()
            return
            
        # Salvar arquivo de áudio
        self._save_audio_file()
        
        logger.info(f"Gravação concluída: {os.path.basename(self.current_audio_file)} ({recording_duration:.0f}ms)")
        
        # Notificar callback
        if self.on_audio_ready and self.current_audio_file:
            try:
                self.on_audio_ready(self.current_audio_file)
            except Exception as e:
                logger.error(f"Erro no callback: {e}")
        
        # Reset
        self.is_recording = False
        self.current_audio_file = None
        self.audio_buffer.clear()

    def _save_audio_file(self):
        """Salvar buffer de áudio como arquivo WAV"""
        if not self.current_audio_file or not self.audio_buffer:
            return
            
        try:
            with wave.open(self.current_audio_file, 'wb') as wav_file:
                wav_file.setnchannels(self.channels)
                wav_file.setsampwidth(2)  # 16-bit
                wav_file.setframerate(self.sample_rate)
                
                # Escrever todos os chunks
                for chunk in self.audio_buffer:
                    wav_file.writeframes(chunk)
                    
        except Exception as e:
            logger.error(f"Erro ao salvar arquivo: {e}")
            self._discard_recording()

    def _discard_recording(self):
        """Descartar gravação atual"""
        if self.current_audio_file and os.path.exists(self.current_audio_file):
            try:
                os.remove(self.current_audio_file)
            except Exception as e:
                logger.error(f"Erro ao remover arquivo: {e}")
                
        self.is_recording = False
        self.current_audio_file = None
        self.audio_buffer.clear()

    def _cleanup_audio(self):
        """Limpar recursos de áudio"""
        try:
            if self.stream:
                self.stream.stop_stream()
                self.stream.close()
                
            if self.audio:
                self.audio.terminate()
                
        except Exception as e:
            logger.error(f"Erro na limpeza de áudio: {e}")

    def stop(self):
        """Parar VAD"""
        if not self.is_listening:
            return
            
        logger.info("Parando VAD...")
        
        # Parar thread
        self.stop_flag.set()
        
        # Finalizar gravação se ativa
        if self.is_recording:
            self._stop_recording()
            
        # Aguardar thread terminar
        if self.recording_thread and self.recording_thread.is_alive():
            self.recording_thread.join(timeout=2)
            
        self.is_listening = False
        logger.info("VAD parado")

    def get_stats(self):
        """Obter estatísticas"""
        return {
            'is_listening': self.is_listening,
            'is_recording': self.is_recording,
            'audio_counter': self.audio_counter,
            'current_file': os.path.basename(self.current_audio_file) if self.current_audio_file else None,
            'buffer_size': len(self.audio_buffer),
            'config': {
                'sample_rate': self.sample_rate,
                'vad_aggressiveness': self.vad_aggressiveness,
                'silence_duration_ms': self.silence_duration_ms,
                'min_recording_duration_ms': self.min_recording_duration_ms
            }
        }

    def test_microphone(self, duration_seconds=3):
        """Testar microfone"""
        logger.info(f"Testando microfone por {duration_seconds} segundos...")
        
        if not self.setup_audio():
            return False
            
        try:
            # Gravar por alguns segundos
            frames = []
            for _ in range(int(self.sample_rate / self.chunk_size * duration_seconds)):
                data = self.stream.read(self.chunk_size)
                frames.append(data)
                
            # Verificar se há sinal
            audio_data = b''.join(frames)
            if len(audio_data) > 0:
                logger.info("✓ Microfone funcionando")
                return True
            else:
                logger.error("✗ Nenhum sinal detectado")
                return False
                
        except Exception as e:
            logger.error(f"✗ Erro no teste: {e}")
            return False
        finally:
            self._cleanup_audio()